### PR TITLE
feature: Facets

### PR DESCRIPTION
**Pull request by coszio**
Needs #758 

Adds `facet` functionality in both remote and local mode

WIP:
- [x] type stubs
- [x] congruence tests

### All Submissions:

* [x] Contributions should target the `dev` branch. Did you create your branch from `dev`?
* [x] Have you followed the guidelines in our Contributing document?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../../pulls) for the same update/change?

### New Feature Submissions:

1. [x] Does your submission pass tests?
2. [x] Have you installed `pre-commit` with `pip3 install pre-commit` and set up hooks with `pre-commit install`?

### Changes to Core Features:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [x] Have you written new tests for your core changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

**Review by joein — CHANGES_REQUESTED** [review]

File: qdrant_client/local/local_collection.py [hunk 1]
```diff
@@ -1071,6 +1090,57 @@ def search_groups(
...
+                facet_hits[v] += 1
+
+        hits = [
+            models.FacetValueHit(value=value, count=count)
+            for value, count in sorted(
+                facet_hits.items(),
+                # order by count descending, then by value ascending
+                key=lambda x: (-x[1], x[0]),
+            )[:limit]
+        ]
```
Comment (joein) [inline qdrant_client/local/local_collection.py:1140]: if there is a facet request on a field which is not one of supported type (int, str), e.g. with a boolean, then we'll get an ugly pydantic error like 

```
validated_self = self.__pydantic_validator__.validate_python(data, self_instance=self)
pydantic_core._pydantic_core.ValidationError: 2 validation errors for FacetValueHit
value.int
  Input should be a valid integer [type=int_type, input_value=True, input_type=bool]
    For further information visit https://errors.pydantic.dev/2.9/v/int_type
value.str
  Input should be a valid string [type=string_type, input_value=True, input_type=bool]
    For further information visit https://errors.pydantic.dev/2.9/v/string_type
```

File: qdrant_client/qdrant_remote.py [hunk 2]
```diff
@@ -1261,7 +1261,7 @@ def recommend_groups(
                 with_lookup = RestToGrpc.convert_with_lookup(with_lookup)
 
             if isinstance(with_lookup, str):
-                with_lookup = grpc.WithLookup(lookup_index=with_lookup)
+                with_lookup = grpc.WithLookup(collection=with_lookup)
```
[the remark this answers is not in the document]
Comment (coszio) [reply qdrant_client/qdrant_remote.py:1264]: 🤷

File: qdrant_client/local/local_collection.py (hunk 1 above)
Comment (coszio) [reply qdrant_client/local/local_collection.py:1137]: I thought this was already sanitized above. Did you create a failing test?

Comment (coszio) [reply qdrant_client/local/local_collection.py:1137]: I just realized that bool is a subclass of int, which is why it was getting through `isinstance`. 

I now fixed it by using

```python
if type(v) not in get_args_subscribed(types.FacetValue):
    continue
```

**Review by joein — APPROVED** [review]

**Comment by coszio** [comment]
Just updated for testing bool facets. Right now it fails because dev build is not yet ready, I'll rerun in a few hours

**Pull request merged**